src/Common/GuardedPoolAllocator.cpp

```diff
@@ -0,0 +1,328 @@
+    return ptr;
+}
+
+AllocationMetadata * GuardedPoolAllocator::addrToMetadata(uintptr_t ptr) const
+{
+    return &metadata[state.getNearestSlot(ptr)];
+}
+
+uintptr_t GuardedPoolAllocator::alignDown(uintptr_t ptr, size_t alignment)
+{
+    assert(isPowerOfTwo(alignment));
+    assert(alignment != 0);
+    if ((ptr & (alignment - 1)) == 0)
+        return ptr;
+
+    ptr -= ptr & (alignment - 1);
+    return ptr;
+}
+
+void GuardedPoolAllocator::init(const clickhouse_gwp_asan::options::Options & options)
+{
+    if (!options.enabled || options.sample_rate == 0)
+        return;
+
+    check(options.sample_rate >= 0, "GWP-ASan Error: sample_rate is < 0.");
+    check(options.sample_rate < (1 << 30), "GWP-ASan Error: sample_rate is >= 2^30.");
+    check(options.max_simultaneous_allocations >= 0, "GWP-ASan Error: max_simultaneous_allocations is < 0.");
+
+    state.max_simultaneous_allocations = options.max_simultaneous_allocations;
+
+    const size_t page_size = getPlatformPageSize();
+    state.page_size = page_size;
+    state.slot_size = page_size * options.slot_size;
+
+    /// check that page size is a power of 2
+    assert(isPowerOfTwo(page_size));
+    size_t pool_size
+        = page_size * (1 + state.max_simultaneous_allocations) + state.max_simultaneous_allocations * state.maximumAllocationSize();
+
+    assert(pool_size % page_size == 0);
+    void * guarded_pool_memory = reserveGuardedPool(pool_size);
+    state.guarded_page_pool = reinterpret_cast<uintptr_t>(guarded_pool_memory);
+    state.guarded_page_pool_end = reinterpret_cast<uintptr_t>(guarded_pool_memory) + pool_size;
+
+    size_t free_slots_required_bytes = roundUpTo(state.max_simultaneous_allocations * sizeof(*free_slots), page_size);
+    assert((free_slots_required_bytes % state.page_size) == 0);
+    free_slots = reinterpret_cast<size_t *>(map(free_slots_required_bytes));
+
+    size_t metadata_required_bytes = roundUpTo(state.max_simultaneous_allocations * sizeof(*metadata), page_size);
+    assert((metadata_required_bytes % state.page_size) == 0);
+    metadata = reinterpret_cast<AllocationMetadata *>(map(metadata_required_bytes));
+
+    /// Multiply the sample rate by 2 to give a good, fast approximation for (1 /
+    /// sample_rate) chance of sampling.
+    if (options.sample_rate != 1)
+        adjusted_sample_rate_plus_one = static_cast<uint32_t>(options.sample_rate) * 2 + 1;
+    else
+        adjusted_sample_rate_plus_one = 2;
+
+    getThreadLocals()->next_sample_counter = (getRandomUint32T() % (adjusted_sample_rate_plus_one - 1)) + 1;
+    allocator_ptr = this;
+}
+
+void GuardedPoolAllocator::lock()
+{
+    getThreadLocals()->recursive_guard = true;
+    /// Lock mutex to lock any in-progress allocations
+    pool_reservation_mutex.lock();
+}
+
+void * GuardedPoolAllocator::allocate(size_t size, size_t alignment)
+{
+    if (size > state.maximumAllocationSize())
+        return nullptr;
+
+    if (state.guarded_page_pool_end == 0)
```

Comment:
How it is possible to get this?

Reply:
Initial design assumed that GWP should be valid even if zero-initialized. To achieve this I set `allocator_ptr` in the beginning of init, so that `getAllocator()` would always return a valid pointer. In this design `shouldSample()` would return `true` as infrequently as possible if GWP is disabled through env variable, however we would eventually land in `allocate`, where we would check. But I could not achieve initialization of GWP before everything else (to be specific: Protobuf initialized something earlier, which resulted in `nullptr` returned by `getAllocator()`.

So eventually I added checks (`getAllocator() != nullptr`) for GWP everywhere and this **check is no longer needed**, so I will remove it.